[TRAFODION-25] First draft of costing code for DELETE.

DaveBirdsall opened the pull request:
This is the first draft of costing code for the Delete operator
in Trafodion. Formerly a cost of 1 was returned for the Delete
operator. With this change, we can expect better plans for DELETE
statements.

Part of the code for the Update operator is in this check-in as
well, but the main part of the Update code remains to be written.

Two CQDs have been added to control execution of this code.

CQD HBASE_DELETE_COSTING 'ON' causes the new Delete costing code
to be executed. A value of 'OFF' causes the old stub logic (a
cost of 1) to be executed instead. The default for this CQD is
'ON', so this code can be exercised.

CQD HBASE_UPDATE_COSTING 'ON' causes partial Update costing code
to be executed. This results in a stub cost object of 1e32 to
be returned. (Actually, this activates a stub that was already
there in the code; this turned out to be useful in discovering
certain bugs, fixed below.) A value of 'OFF' causes earlier 
stub logic (a cost of 1) to be executed instead. The default
for this CQD is 'OFF' for now.

Once this code has been sufficiently validated, these CQDs will be
removed.

Several bugs are also fixed in this check-in. All are latent,
existing bugs that became more visible when regression-testing the
new costing code due to plan changes.

1. MERGE with a NOT MATCHED action was being (incorrectly)
transformed into a nested join in some plans. This caused the
NOT MATCHED action to not be executed. (optimizer/TransRule.cpp,
optimizer/RelUpdate.h)

2. MERGE error messages generated by the compiler were sometimes
incorrect. (optimizer/BindRelExpr.cpp, generator/GenPreCode.cpp)

3. A check constraint error message was not being reported when
it should have on a table with SERIALIZED columns. This was caused
by an HBase filter expression being generated with operators
having the wrong sense (e.g., < instead of >).
(generator/GenPreCode.cpp)

4. CLEANUP would fail if the record in the OBJECTS_UNIQ_IDX index
was missing. This happened because it was using a vanilla DELETE
and the plan for the DELETE changed with the costing changes to 
use index access. The fix was to use CQDs to avoid accessing the
index when deleting from the OBJECTS table.
(sqlcomp/CmpSeabaseDDLcleanup.cpp)

5. INITIALIZE AUTHORIZATION would fail if there was a row in
the OBJECTS table with an object_owner that does not exist in
the AUTHS table. (The latter can happen in rare circumstances
when toggling authorization off and on repeatedly. That is a
bug in its own right that remains outstanding.) This has been
worked around by writing 'DB__ROOT' to the OBJECT_PRIVILEGES
table when the object_owner is invalid.

Some other issues were noticed during unit testing that will
be addressed in later check-ins:

1. Scan costs for DELETEs using TSJ plans are grossly over-
estimated. The cause is that the generated join predicates between
the Scan and Delete operators are being assigned a default
selectivity of 1/3. This is an existing, latent bug.

2. The HBaseDeleteRule in the optimizer considers a Delete-only
plan (that is, without a TSJ of a Scan to a Delete) when there
are equality predicates on all key columns, and when there are
no predicates on key columns at all. The former is correct, the
latter is almost never a good plan. Costing should prevent the
latter from being chosen but still this is unnecessary extra
computation. (The rule does not consider Delete-only plans when
a proper subset of the key columns have equality predicates.)
This is an existing, latent bug.

Traf-Jenkins commented:
Test Passed.
https://jenkins.esgyn.com/job/Check-PR/186/

zellerh commented on core/sql/optimizer/ScmCostMethod.cpp:
```diff
@@ -3632,6 +3632,561 @@ CostMethodMergeUnion::scmComputeOperatorCostInternal(RelExpr *op,
+    // The char inputs are still in terms of the source. Map them to the target.
+    // Note: The source char outputs in the ipp have already been mapped to
+    // the target. CharOutputs are a set, meaning they do not have duplicates
+    // But we could have cases where two columns of the target are matched to the
```
The updateToSelectMap() is really just a helper structure and its contents should not be examined to make these kind of decisions. I would suggest deleting anything that references bottomValues, bottomValuesSet and useListInsteadOfSet and to do the rewrite that's shown on line 3960. Sorry if I missed some complication that this code is meant to address.

zellerh commented on core/sql/optimizer/ScmCostMethod.cpp:
```diff
@@ -3632,6 +3632,561 @@ CostMethodMergeUnion::scmComputeOperatorCostInternal(RelExpr *op,
+        // it means the table was not partitioned at all, so we don't
+        // need to limit anything since there already is no asynch I/O.
+
+     // TODO: lppf is always null in Trafodion; figure out what to do instead...
```
I think this if (lines 3999 to 4011) can be deleted. countOfAsynchronousStreams is already set for the regular case.

zellerh commented on core/sql/optimizer/ScmCostMethod.cpp:
```diff
@@ -3632,6 +3632,561 @@ CostMethodMergeUnion::scmComputeOperatorCostInternal(RelExpr *op,
+    }
+  }
+
+  // Notes: At execution time, several different TCBs can be created
```
Thanks, very useful comment!

zellerh commented:
Looks great to me, with two small comments.

DaveBirdsall commented:
Yes, the section of code you mention, Hans, needs more work. That particular code is inherited from Seaquest. It is the part of the code that I understand least well. I will be unit testing it carefully (and perhaps throwing it away and rewriting it) on the next submittal.

sureshsubbiah commented on core/sql/generator/GenPreCode.cpp:
```diff
@@ -4385,12 +4377,14 @@ RelExpr * GenericUpdate::preCodeGen(Generator * generator,
 	}
 
-      if ((getOperatorType() == REL_HBASE_DELETE) &&
+      if ((getOperatorType() != REL_HBASE_UPDATE) &&
```
Just curious. What value can getOperatorType() have at this point in code other than REL_HBASE_DELETE and REL_HBASE_UPDATE?

DaveBirdsall replied:
I think that's it.

sureshsubbiah commented on core/sql/generator/GenPreCode.cpp:
```diff
@@ -4368,14 +4368,6 @@ RelExpr * GenericUpdate::preCodeGen(Generator * generator,
-			      << DgString0(" Non-unique ON clause not allowed with INSERT.");
-	  GenExit();
-	}
-
```
I am glad that this check was removed. Can you please tell me if we ever see any of the other 3241 errors being triggered from the generator? To me this indicates that the binder "let one get through". I would like to eventually raise all these 3241 errors in thebinder.

DaveBirdsall replied:
Yes, some of the 3241 errors remain in the generator. I had to move this particular test up to the binder because with the costing changes I was sometimes failing to get a plan in phase 1 of optimization.

sureshsubbiah commented on core/sql/optimizer/BindRelExpr.cpp:
```diff
@@ -11046,6 +11046,14 @@ RelExpr *MergeDelete::bindNode(BindWA *bindWA)
         bindWA->setErrStatus();
         return this;
       }
+      if (CmpCommon::getDefault(COMP_BOOL_175) == DF_OFF)
```
Does the optimizer fail to produce a plan with this change or even before? Maybe a JIRA or a comment stating the failing DELETE+INSERT merge queries will be useful. We may discussed this in the dev thread before, but is it possible that UPDATE+INSERT could have a similar issue? UPDATE+INSERT MERGE statements are now used more than before since any UPSERT statement with INDEX gets transformed to them.

DaveBirdsall replied:
UPDATE + INSERT does not have the issue; I saw successful TSJ plans with UPDATE + INSERT. With DELETE + INSERT, the TSJ plans fail. DELETE + INSERT was turned off by default before; only by turning on COMP_BOOL_175 to ON would it be allowed. I took this to mean that it was known to be buggy, and left it in this state. I spent about a half day debugging it without success and abandoned that effort for now as it seemed outside the scope of the costing work. I will file a JIRA.

DaveBirdsall replied:
I filed JIRA TRAFODION-1509, and noted it in the comments in the code below.

sureshsubbiah commented on core/sql/optimizer/ScmCostMethod.cpp:
```diff
@@ -3632,6 +3632,561 @@ CostMethodMergeUnion::scmComputeOperatorCostInternal(RelExpr *op,
+  NABoolean probesInOrder
+  ) const
+{
+  const CostScalar indexLevels = CIDesc->getIndexLevels();
```
What does indexLevels mean for HBase get or scan ?

nonstop-qfchen replied:
IndexLevels is the number of index levels in a randomly selected HFile, excluding the data blocks.  See Java method FixedFileTrailer::getNumDataIndexLevels(). The JNI code for it is getHbaseTableInfo().

DaveBirdsall replied:
Good catch. I'll remove this line of code.

sureshsubbiah commented on core/sql/optimizer/TransRule.cpp:
```diff
@@ -3287,6 +3287,15 @@ NABoolean TSJRule::topMatch (RelExpr * expr,
+  // row would not be returned by the outer child scan node, so
+  // the inner child merge node would never see it and hence the
+  // "NOT MATCHED" logic would not be activiated.)
+  if (updateExpr->isMerge() && updateExpr->insertValues())
```
I am assuming that when a TSJ is used for IndexMaintenence or to to flow rows from a TupleList or a source select it should be acceptable. Are those cases excluded somehow by this restriction?

DaveBirdsall replied:
Yes, a TSJ for index maintenance still occurs. That is introduced by the Binder, before we get to this code. This code deals with a different issue. When we parse an UPDATE or DELETE statement, the Parser creates a Scan node underneath the Update or Delete node. This rule allows the optimizer to transform that to a TSJ(Scan,Update/Delete). (The alternative plan simply clips off the Scan and an Update/Delete node by itself remains.) For MERGE with a NOT MATCHED clause, the TSJ transformation gives an incorrect plan.

nonstop-qfchen commented on core/sql/generator/GenPreCode.cpp:
```diff
@@ -4385,12 +4377,14 @@ RelExpr * GenericUpdate::preCodeGen(Generator * generator,
+          // the CQD is on
 	  *CmpCommon::diags() << DgSqlCode(-3241)
-			      << DgString0(" MERGE delete not allowed with INSERT.");
+			      << DgString0(" This MERGE is not allowed with INSERT.");
```
Maybe reword the error message to "This MERGE with DELETE is not allowed with INSERT on non-Trafodion tables".

DaveBirdsall replied:
OK.

DaveBirdsall replied:
Actually, the message is correct as-is. It is preventing MERGE DELETE + INSERT on Trafodion tables as well as non-Trafodion tables.

nonstop-qfchen commented on core/sql/generator/GenPreCode.cpp:
```diff
@@ -11180,13 +11179,33 @@ NABoolean HbaseAccess::isHbaseFilterPred(Generator * generator, ItemExpr * ie,
+            if (flipOp)
+              op = "GREATER_OR_EQUAL";
+            else
+	      op = "LESS_OR_EQUAL";
```
should be "LESS".

nonstop-qfchen commented on core/sql/optimizer/BindRelExpr.cpp:
```diff
@@ -11046,6 +11046,14 @@ RelExpr *MergeDelete::bindNode(BindWA *bindWA)
       }
+      if (CmpCommon::getDefault(COMP_BOOL_175) == DF_OFF)
+      {
+        // MERGE DELETE + INSERT is buggy, so disallow it unless CQD is on. In
```
Since COMP_BOOL_175 is OFF by default, this check prevents MERGE_DELETE + INSERT.  Is it intended?

DaveBirdsall replied:
Yes.

DaveBirdsall replied:
I've written a new JIRA, TRAFODION-1509, to explain this situation, and will note it in the commends here.

nonstop-qfchen commented on core/sql/optimizer/CostMethod.h:
```diff
@@ -2036,30 +2037,31 @@ class CostMethodHbaseInsert : public CostMethod
 					       Lng32& countOfStreams) = 0;
 
-  // Common functions to cost DP2 operations.
+  // Common functions to cost Hbase operations.
```
Maybe remove the function declaration since it has not been implemented in Trafodion.

DaveBirdsall replied:
Will try to wordsmith this comment in a more meaningful way.

nonstop-qfchen commented on core/sql/optimizer/OptPhysRelExpr.cpp:
```diff
@@ -16546,6 +16546,18 @@ RelInternalSP::costMethod() const
+HbaseDelete::costMethod() const
+{
+  if (CmpCommon::getDefault(HBASE_DELETE_COSTING) == DF_OFF)
+    return RelExpr::costMethod();  // returns cost 1 cost object
```
maybe say "returns the default cost method that returns an object of cost 1".

DaveBirdsall replied:
Will reword.

nonstop-qfchen commented on core/sql/optimizer/ScmCostMethod.cpp:
```diff
@@ -3632,6 +3632,561 @@ CostMethodMergeUnion::scmComputeOperatorCostInternal(RelExpr *op,
+
+  } // if there are key predicates
+
+  // If there is no key predicates, a full table scan will be generated.
```
If we do not take the existence of key predicates path, we never apply the selHint or cardHint. Suggest to move the call  @3761 out of the block so that it can be called when there is no key predicate case.

DaveBirdsall replied:
Will attempt to do this.

DaveBirdsall replied:
This is now done, with the rework.

nonstop-qfchen commented on core/sql/optimizer/ScmCostMethod.cpp:
```diff
@@ -3632,6 +3632,561 @@ CostMethodMergeUnion::scmComputeOperatorCostInternal(RelExpr *op,
+	activePartitions,
+	CIDesc
+	);
+    }
```
We may need to add a low bound for numRowsToScan to be at least numRowsToDelete, when the stats is available.

DaveBirdsall replied:
Good point. I will add this.

nonstop-qfchen commented on core/sql/optimizer/ScmCostMethod.cpp:
```diff
@@ -3632,6 +3632,561 @@ CostMethodMergeUnion::scmComputeOperatorCostInternal(RelExpr *op,
+  CostScalar randomIOs(csZero);
+  CostScalar sequentialIOs(csZero);
+
+  CostScalar countOfAsynchronousStreams = activePartitions;
```
should we assert that getProbesForceSynchronousAccessFlag() is FALSE here?

DaveBirdsall replied:
I will investigate.

nonstop-qfchen commented on core/sql/optimizer/ScmCostMethod.cpp:
```diff
@@ -3632,6 +3632,561 @@ CostMethodMergeUnion::scmComputeOperatorCostInternal(RelExpr *op,
+
+  tuplesProduced = numRowsToDelete;
+  tuplesProcessed = numRowsToScan; 
+  tuplesSent = numRowsToDelete;
```
I wonder if we even care tupleSent since the DELETE operator never sends rows to another Trafodion process through ESP.

DaveBirdsall replied:
The tuplesSent here is meant to model message cost between HBase client and server. I followed the model of the ScanOptimizer here, which does something similar.

nonstop-qfchen commented on core/sql/optimizer/ScmCostMethod.cpp:
```diff
@@ -3632,6 +3632,561 @@ CostMethodMergeUnion::scmComputeOperatorCostInternal(RelExpr *op,
+  CostScalar outputRowSize = delOp->getGroupAttr()->getRecordLength();
+  CostScalar outputRowSizeFactor = scmRowSizeFactor(outputRowSize);
+
+  tuplesProcessed = tuplesProcessed * rowSizeFactor;
```
Could use tupleProcessed *= rowSizeFactor;

DaveBirdsall replied:
OK.

nonstop-qfchen commented on core/sql/regress/seabase/EXPECTED011:
```diff
@@ -299,8 +299,10 @@ LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
 
-1    .    2    root                  o         x                     2.00E+000
-.    .    1    trafodion_delete                T011T1                2.00E+000
+3    .    4    root                            x                     2.00E+000
```
I was surprised to see plan change for deleting very few number of rows.  Sounds like we need to run some performance tests?

DaveBirdsall replied:
Definitely! This is still a work in progress.

robertamarton commented on core/sql/sqlcomp/PrivMgrPrivileges.cpp:
```diff
@@ -5390,7 +5390,8 @@ PrivStatus ObjectPrivsMDTable::updateWhere(const std::string & setClause,
 //    object_owner, -- granteeID
-//    (select auth_db_name from AUTHS where auth_id = object_owner), --granteeName
+//    coalesce((select auth_db_name from AUTHS where auth_id = object_owner),
+//             'DB__ROOT') --granteeName
```
I wonder what would happen if the granteeName becomes DB__ROOT but the granteeID is not DB__ROOT ID (33333) but something else.

DaveBirdsall replied:
I will investigate. If you prefer, I can use something different than DB__ROOT. I tried DB__ROOT as it seemed likely to be the most restrictive.

robertamarton replied:
This is probably okay.  I am just wondering what happens if you initialize authorization, create tables with role and user owners other than DB__ROOT, then drop authorization.  Not sure if we reset object ownership.  Later when we initialize authorization, we may be missing things.   Some testing is needed and perhaps a new JIRA created.  Unfortunately I don't have a working version so I can't test.  I will test is out when I do and create a JIRA, if needed.

sureshsubbiah commented:
Dave, should we merge this now or would you add more changes to this delivery ?

DaveBirdsall commented:
I will be pushing some reworks up shortly (as soon as a regression run finishes cleanly). Please hold off on merge.

Traf-Jenkins commented:
Test Failed.
https://jenkins.esgyn.com/job/Check-PR/195/

DaveBirdsall commented:
jenkins, retest

Traf-Jenkins commented:
Test Passed.
https://jenkins.esgyn.com/job/Check-PR/196/

DaveBirdsall commented:
Reviewers, any final comments on the reworked code before committing?

nonstop-qfchen commented:
Hi Dave,

Can you please send us the URL to the rework?

Thank you. --Qifan

On Wed, Oct 7, 2015 at 1:28 PM, DaveBirdsall <notifications@github.com>
wrote:

> Reviewers, any final comments on the reworked code before committing?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-trafodion/pull/102#issuecomment-146285744>
> .
>



-- 
Regards, --Qifan

DaveBirdsall commented:
The rework can be found with the pull request, here: https://github.com/apache/incubator-trafodion/pull/102. In the Conversation tab, look for my comment, "I will be pushing up some reworks shortly...". Right below that in tiny print you'll see a commit, "Reworks addressing review comments". Click on that.

nonstop-qfchen commented:
Hi Dave,

The rework looks good.

I wonder if we should set HBASE_DELETE_COSTING to off by default, and turn
it back on after the performance test.

Thanks -Qifan

On Wed, Oct 7, 2015 at 1:49 PM, DaveBirdsall <notifications@github.com>
wrote:

> The rework can be found with the pull request, here: #102
> <https://github.com/apache/incubator-trafodion/pull/102>. In the
> Conversation tab, look for my comment, "I will be pushing up some reworks
> shortly...". Right below that in tiny print you'll see a commit, "Reworks
> addressing review comments". Click on that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-trafodion/pull/102#issuecomment-146291541>
> .
>



-- 
Regards, --Qifan

DaveBirdsall commented:
I would prefer to leave HBASE_DELETE_COSTING on, as the off behavior stubs delete cost as 1 for every delete operation (which is obviously very wrong). Early performance testing is yielding good results; I'm seeing good switch over from index access scan on a delete to full table scan as the selectivity of the index predicate rises for example.

nonstop-qfchen commented:
Yes, if we can document evidences of the good plans now being generated in
this email thread, that will be sufficient to leave the CQD on.

We also will need to arrange the performance test.  I'll ask about it,
along with other items on the plate.

Thanks.

On Wed, Oct 7, 2015 at 2:12 PM, DaveBirdsall <notifications@github.com>
wrote:

> I would prefer to leave HBASE_DELETE_COSTING on, as the off behavior stubs
> delete cost as 1 for every delete operation (which is obviously very
> wrong). Early performance testing is yielding good results; I'm seeing good
> switch over from index access scan on a delete to full table scan as the
> selectivity of the index predicate rises for example.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-trafodion/pull/102#issuecomment-146298929>
> .
>



-- 
Regards, --Qifan

DaveBirdsall commented:
Will publish some test results shortly.

DaveBirdsall commented:
Below is a log of a test of the DELETE costing changes. In this particular test, I'm looking at plan quality. I also attempt some performance tests at the end, though the results here were done on a VM with uncontrolled competing workload and so must be taken with a grain of salt. Some interesting things to note. Near the end of the test, I vary the cardinality of the set of rows being deleted to see when we switch over from using an index scan to a base table scan when deciding which rows to delete. The crossover point is between queries D7000 and D7500. I also execute these DELETEs and capture their execution time statistics. The plan execution times are in the same ballpark, which suggests that the crossover point is reasonable. I was watching with "top" while these DELETEs were running; during the first one there were two competing builds running on the system (hence the higher elapsed time), for the next four only one competing build. Builds vary in their resource use over time so this affected the test times.

>>obey TEST001(setup1);
>>
>>drop table if exists np001;

--- SQL operation complete.
>>drop table if exists np001backup;

--- SQL operation complete.
>>
>>-- create a non-partitioned table with 16384 rows in it
>>
>>create table np001
+> ( a integer not null,
+>   b integer not null,
+>   c integer not null,
+>   d varchar(100),
+>   primary key (a) );

--- SQL operation complete.
>>
>>
>>insert into np001 values (1,1,1,'hi there!'),
+>  (2,1,1,'a somewhat longer string'),
+>  (3,2,2,'Now is the time for all good men to come to the aid of their country.'),
+>  (4,3,3,null);

--- 4 row(s) inserted.
>>
>>insert into np001 select a+4,b+4,c,d from np001;

--- 4 row(s) inserted.
>>
>>insert into np001 select a+8,b+8,c,d from np001;

--- 8 row(s) inserted.
>>
>>insert into np001 select a+16,b+16,c,d from np001;

--- 16 row(s) inserted.
>>
>>insert into np001 select a+32,b+32,c,d from np001;

--- 32 row(s) inserted.
>>
>>insert into np001 select a+64,b+64,c,d from np001;

--- 64 row(s) inserted.
>>
>>insert into np001 select a+128,b+128,c,d from np001;

--- 128 row(s) inserted.
>>
>>insert into np001 select a+256,b+256,c,d from np001;

--- 256 row(s) inserted.
>>
>>insert into np001 select a+512,b+512,c,d from np001;

--- 512 row(s) inserted.
>>
>>upsert into np001 select a+1024,b+1024,c,d from np001;

--- 1024 row(s) inserted.
>>
>>upsert into np001 select a+2048,b+2048,c,d from np001;

--- 2048 row(s) inserted.
>>
>>upsert into np001 select a+4096,b+4096,c,d from np001;

--- 4096 row(s) inserted.
>>
>>upsert into np001 select a+8192,b+8192,c,d from np001;

--- 8192 row(s) inserted.
>>
>>-- turn off histogram caching so we'll get the right histograms afterward
>>cqd cache_histograms 'OFF';

--- SQL operation complete.
>>
>>update statistics for table np001 on every column;

--- SQL operation complete.
>>
>>cqd cache_histograms reset;

--- SQL operation complete.
>>
>>-- create a copy of the data so we can put it back after deletes
>>create table np001backup like np001;

--- SQL operation complete.
>>
>>insert into np001backup select * from np001;

--- 16384 row(s) inserted.
>>
>>obey TEST001(plans);
>>
>>-- turn off query caching
>>-- so we can see effects of particular selectivities
>>cqd query_cache '0';

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where a = 1025;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                  o                               1.00E+000
.    .    1    trafodion_delete                NP001                 1.00E+000

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where a between 1024 and 1034;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

3    .    4    root                            x                     1.20E+001
1    2    3    tuple_flow                                            1.20E+001
.    .    2    trafodion_vsbb_delet            NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 1.20E+001

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where a between 1024 and 15536;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

3    .    4    root                            x                     1.45E+004
1    2    3    tuple_flow                                            1.45E+004
.    .    2    trafodion_vsbb_delet            NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 1.45E+004

--- SQL operation complete.
>>
>>explain options 'f' delete from np001;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

3    .    4    root                            x                     1.63E+004
1    2    3    tuple_flow                                            1.63E+004
.    .    2    trafodion_vsbb_delet            NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 1.63E+004

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where b = 1025;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

3    .    4    root                            x                     1.00E+000
1    2    3    tuple_flow                                            1.00E+000
.    .    2    trafodion_vsbb_delet            NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 1.00E+000

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where b between 1024 and 1034;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

3    .    4    root                            x                     1.29E+001
1    2    3    tuple_flow                                            1.29E+001
.    .    2    trafodion_vsbb_delet            NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 1.29E+001

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where b between 1024 and 15536;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

3    .    4    root                            x                     1.45E+004
1    2    3    tuple_flow                                            1.45E+004
.    .    2    trafodion_vsbb_delet            NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 1.45E+004

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where c = 1;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

3    .    4    root                            x                     8.19E+003
1    2    3    tuple_flow                                            8.19E+003
.    .    2    trafodion_vsbb_delet            NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 8.19E+003

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where c = 2;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

3    .    4    root                            x                     4.09E+003
1    2    3    tuple_flow                                            4.09E+003
.    .    2    trafodion_vsbb_delet            NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 4.09E+003

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where c >= 1 and c <= 2;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

3    .    4    root                            x                     1.22E+004
1    2    3    tuple_flow                                            1.22E+004
.    .    2    trafodion_vsbb_delet            NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 1.22E+004

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where d = 'non-existent row';

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

3    .    4    root                            x                     4.09E+003
1    2    3    tuple_flow                                            4.09E+003
.    .    2    trafodion_vsbb_delet            NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 4.09E+003

--- SQL operation complete.
>>
>>-- turn query caching back on
>>cqd query_cache reset;

--- SQL operation complete.
>>
>>obey TEST001(planssel);
>>
>>-- select plans just for comparison with deletes
>>
>>-- turn off query caching
>>-- so we can see effects of particular selectivities
>>cqd query_cache '0';

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where a = 1025;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                  o                               1.00E+000
.    .    1    trafodion_scan                  NP001                 1.00E+000

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where a between 1024 and 1034;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  1.20E+001
.    .    1    trafodion_scan                  NP001                 1.20E+001

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where a between 1024 and 15536;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  1.45E+004
.    .    1    trafodion_scan                  NP001                 1.45E+004

--- SQL operation complete.
>>
>>explain options 'f' select a from np001;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  1.63E+004
.    .    1    trafodion_scan                  NP001                 1.63E+004

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where b = 1025;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  1.00E+000
.    .    1    trafodion_scan                  NP001                 1.00E+000

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where b between 1024 and 1034;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  1.29E+001
.    .    1    trafodion_scan                  NP001                 1.29E+001

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where b between 1024 and 15536;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  1.45E+004
.    .    1    trafodion_scan                  NP001                 1.45E+004

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where c = 1;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  8.19E+003
.    .    1    trafodion_scan                  NP001                 8.19E+003

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where c = 2;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  4.09E+003
.    .    1    trafodion_scan                  NP001                 4.09E+003

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where c >= 1 and c <= 2;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  1.22E+004
.    .    1    trafodion_scan                  NP001                 1.22E+004

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where d = 'non-existent row';

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  4.09E+003
.    .    1    trafodion_scan                  NP001                 4.09E+003

--- SQL operation complete.
>>
>>-- turn query caching back on
>>cqd query_cache reset;

--- SQL operation complete.
>>
>>obey TEST001(setup2);
>>
>>-- index on a high UEC column
>>create index inp001 on np001(b);

--- SQL operation complete.
>>
>>-- index on a low UEC column
>>create index inp002 on np001(c);

--- SQL operation complete.
>>
>>-- turn off histogram caching so we'll get the right histograms afterward
>>cqd cache_histograms 'OFF';

--- SQL operation complete.
>>
>>update statistics for table np001 on every column;

--- SQL operation complete.
>>
>>cqd cache_histograms reset;

--- SQL operation complete.
>>
>>obey TEST001(plans);
>>
>>-- turn off query caching
>>-- so we can see effects of particular selectivities
>>cqd query_cache '0';

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where a = 1025;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

5    .    6    root                  o         x                     2.00E+000
1    4    5    nested_join                                           2.00E+000
2    3    4    merge_union                                           2.00E+000
.    .    3    trafodion_delete                INP002                1.00E+000
.    .    2    trafodion_delete                INP001                1.00E+000
.    .    1    trafodion_delete                NP001                 1.00E+000

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where a between 1024 and 1034;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

7    .    8    root                            x                     2.40E+001
3    6    7    nested_join                                           2.40E+001
4    5    6    merge_union                                           2.00E+000
.    .    5    trafodion_vsbb_delet            INP002                1.00E+000
.    .    4    trafodion_vsbb_delet            INP001                1.00E+000
1    2    3    nested_join                                           1.20E+001
.    .    2    trafodion_delete                NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 1.20E+001

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where a between 1024 and 15536;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

7    .    8    root                            x                     2.90E+004
3    6    7    nested_join                                           2.90E+004
4    5    6    merge_union                                           2.00E+000
.    .    5    trafodion_vsbb_delet            INP002                1.00E+000
.    .    4    trafodion_vsbb_delet            INP001                1.00E+000
1    2    3    nested_join                                           1.45E+004
.    .    2    trafodion_delete                NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 1.45E+004

--- SQL operation complete.
>>
>>explain options 'f' delete from np001;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

7    .    8    root                            x                     3.27E+004
3    6    7    nested_join                                           3.27E+004
4    5    6    merge_union                                           2.00E+000
.    .    5    trafodion_vsbb_delet            INP002                1.00E+000
.    .    4    trafodion_vsbb_delet            INP001                1.00E+000
1    2    3    nested_join                                           1.63E+004
.    .    2    trafodion_delete                NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 1.63E+004

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where b = 1025;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

10   .    11   root                            x                     2.00E+000
6    9    10   nested_join                                           2.00E+000
7    8    9    merge_union                                           2.00E+000
.    .    8    trafodion_vsbb_delet            INP002                1.00E+000
.    .    7    trafodion_vsbb_delet            INP001                1.00E+000
4    5    6    nested_join                                           1.00E+000
.    .    5    trafodion_delete                NP001                 1.00E+000
1    3    4    nested_join                                           1.00E+000
2    .    3    probe_cache                                           1.00E+000
.    .    2    trafodion_vsbb_scan             NP001                 1.00E+000
.    .    1    trafodion_index_scan            INP001                1.00E+000

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where b between 1024 and 1034;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

10   .    11   root                            x                     2.59E+001
6    9    10   nested_join                                           2.59E+001
7    8    9    merge_union                                           2.00E+000
.    .    8    trafodion_vsbb_delet            INP002                1.00E+000
.    .    7    trafodion_vsbb_delet            INP001                1.00E+000
4    5    6    nested_join                                           1.29E+001
.    .    5    trafodion_delete                NP001                 1.00E+000
1    3    4    nested_join                                           1.29E+001
2    .    3    probe_cache                                           1.00E+000
.    .    2    trafodion_vsbb_scan             NP001                 1.00E+000
.    .    1    trafodion_index_scan            INP001                1.29E+001

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where b between 1024 and 15536;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

7    .    8    root                            x                     2.90E+004
3    6    7    nested_join                                           2.90E+004
4    5    6    merge_union                                           2.00E+000
.    .    5    trafodion_vsbb_delet            INP002                1.00E+000
.    .    4    trafodion_vsbb_delet            INP001                1.00E+000
1    2    3    nested_join                                           1.45E+004
.    .    2    trafodion_delete                NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 1.45E+004

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where c = 1;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

10   .    11   root                            x                     1.63E+004
6    9    10   nested_join                                           1.63E+004
7    8    9    merge_union                                           2.00E+000
.    .    8    trafodion_vsbb_delet            INP002                1.00E+000
.    .    7    trafodion_vsbb_delet            INP001                1.00E+000
4    5    6    nested_join                                           8.19E+003
.    .    5    trafodion_delete                NP001                 1.00E+000
1    3    4    nested_join                                           8.19E+003
2    .    3    probe_cache                                           5.00E-001
.    .    2    trafodion_vsbb_scan             NP001                 5.00E-001
.    .    1    trafodion_index_scan            INP002                8.19E+003

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where c = 2;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

10   .    11   root                            x                     8.19E+003
6    9    10   nested_join                                           8.19E+003
7    8    9    merge_union                                           2.00E+000
.    .    8    trafodion_vsbb_delet            INP002                1.00E+000
.    .    7    trafodion_vsbb_delet            INP001                1.00E+000
4    5    6    nested_join                                           4.09E+003
.    .    5    trafodion_delete                NP001                 1.00E+000
1    3    4    nested_join                                           4.09E+003
2    .    3    probe_cache                                           2.50E-001
.    .    2    trafodion_vsbb_scan             NP001                 2.50E-001
.    .    1    trafodion_index_scan            INP002                4.09E+003

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where c >= 1 and c <= 2;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

7    .    8    root                            x                     2.45E+004
3    6    7    nested_join                                           2.45E+004
4    5    6    merge_union                                           2.00E+000
.    .    5    trafodion_vsbb_delet            INP002                1.00E+000
.    .    4    trafodion_vsbb_delet            INP001                1.00E+000
1    2    3    nested_join                                           1.22E+004
.    .    2    trafodion_delete                NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 1.22E+004

--- SQL operation complete.
>>
>>explain options 'f' delete from np001 where d = 'non-existent row';

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

7    .    8    root                            x                     8.19E+003
3    6    7    nested_join                                           8.19E+003
4    5    6    merge_union                                           2.00E+000
.    .    5    trafodion_vsbb_delet            INP002                1.00E+000
.    .    4    trafodion_vsbb_delet            INP001                1.00E+000
1    2    3    nested_join                                           4.09E+003
.    .    2    trafodion_delete                NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 4.09E+003

--- SQL operation complete.
>>
>>-- turn query caching back on
>>cqd query_cache reset;

--- SQL operation complete.
>>
>>obey TEST001(planssel);
>>
>>-- select plans just for comparison with deletes
>>
>>-- turn off query caching
>>-- so we can see effects of particular selectivities
>>cqd query_cache '0';

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where a = 1025;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                  o                               1.00E+000
.    .    1    trafodion_scan                  NP001                 1.00E+000

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where a between 1024 and 1034;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  1.20E+001
.    .    1    trafodion_scan                  NP001                 1.20E+001

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where a between 1024 and 15536;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  1.45E+004
.    .    1    trafodion_index_scan            INP002                1.45E+004

--- SQL operation complete.
>>
>>explain options 'f' select a from np001;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  1.63E+004
.    .    1    trafodion_index_scan            INP001                1.63E+004

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where b = 1025;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  1.00E+000
.    .    1    trafodion_index_scan            INP001                1.00E+000

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where b between 1024 and 1034;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  1.29E+001
.    .    1    trafodion_index_scan            INP001                1.29E+001

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where b between 1024 and 15536;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  1.45E+004
.    .    1    trafodion_index_scan            INP001                1.45E+004

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where c = 1;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  8.19E+003
.    .    1    trafodion_index_scan            INP002                8.19E+003

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where c = 2;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  4.09E+003
.    .    1    trafodion_index_scan            INP002                4.09E+003

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where c >= 1 and c <= 2;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  1.22E+004
.    .    1    trafodion_index_scan            INP002                1.22E+004

--- SQL operation complete.
>>
>>explain options 'f' select a from np001 where d = 'non-existent row';

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

1    .    2    root                                                  4.09E+003
.    .    1    trafodion_scan                  NP001                 4.09E+003

--- SQL operation complete.
>>
>>-- turn query caching back on
>>cqd query_cache reset;

--- SQL operation complete.
>>
>>obey TEST001(crossover);
>>
>>-- perform some DELETEs and measure the time taken, compare with
>>-- the plans generated
>>
>>-- what we see is that as the cardinality increases, we switch from
>>-- doing index access scan to a full table scan
>>
>>-- we will also see that the execution times are about the same for
>>-- both plans in the neighborhood of the cross over; however this has
>>-- to be taken with a grain of salt as these test results were done
>>-- on a VM with no controls over competing activity
>>
>>prepare D6000 from delete from np001 where b between 1000 and 6000;

--- SQL command prepared.
>>prepare D6500 from delete from np001 where b between 1000 and 6500;

--- SQL command prepared.
>>prepare D7000 from delete from np001 where b between 1000 and 7000;

--- SQL command prepared.
>>prepare D7500 from delete from np001 where b between 1000 and 7500;

--- SQL command prepared.
>>prepare D8000 from delete from np001 where b between 1000 and 8000;

--- SQL command prepared.
>>
>>explain options 'f' D6000;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

11   .    12   root                            x                     1.00E+004
7    10   11   nested_join                                           1.00E+004
8    9    10   merge_union                                           2.00E+000
.    .    9    trafodion_vsbb_delet            INP002                1.00E+000
.    .    8    trafodion_vsbb_delet            INP001                1.00E+000
5    6    7    nested_join                                           5.00E+003
.    .    6    trafodion_delete                NP001                 1.00E+000
4    .    5    sort                                                  5.00E+003
1    3    4    nested_join                                           5.00E+003
2    .    3    probe_cache                                           1.00E+000
.    .    2    trafodion_vsbb_scan             NP001                 1.00E+000
.    .    1    trafodion_index_scan            INP001                5.00E+003

--- SQL operation complete.
>>explain options 'f' D6500;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

11   .    12   root                            x                     1.10E+004
7    10   11   nested_join                                           1.10E+004
8    9    10   merge_union                                           2.00E+000
.    .    9    trafodion_vsbb_delet            INP002                1.00E+000
.    .    8    trafodion_vsbb_delet            INP001                1.00E+000
5    6    7    nested_join                                           5.50E+003
.    .    6    trafodion_delete                NP001                 1.00E+000
4    .    5    sort                                                  5.50E+003
1    3    4    nested_join                                           5.50E+003
2    .    3    probe_cache                                           1.00E+000
.    .    2    trafodion_vsbb_scan             NP001                 1.00E+000
.    .    1    trafodion_index_scan            INP001                5.50E+003

--- SQL operation complete.
>>explain options 'f' D7000;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

11   .    12   root                            x                     1.20E+004
7    10   11   nested_join                                           1.20E+004
8    9    10   merge_union                                           2.00E+000
.    .    9    trafodion_vsbb_delet            INP002                1.00E+000
.    .    8    trafodion_vsbb_delet            INP001                1.00E+000
5    6    7    nested_join                                           6.00E+003
.    .    6    trafodion_delete                NP001                 1.00E+000
4    .    5    sort                                                  6.00E+003
1    3    4    nested_join                                           6.00E+003
2    .    3    probe_cache                                           1.00E+000
.    .    2    trafodion_vsbb_scan             NP001                 1.00E+000
.    .    1    trafodion_index_scan            INP001                6.00E+003

--- SQL operation complete.
>>explain options 'f' D7500;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

7    .    8    root                            x                     1.30E+004
3    6    7    nested_join                                           1.30E+004
4    5    6    merge_union                                           2.00E+000
.    .    5    trafodion_vsbb_delet            INP002                1.00E+000
.    .    4    trafodion_vsbb_delet            INP001                1.00E+000
1    2    3    nested_join                                           6.50E+003
.    .    2    trafodion_delete                NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 6.50E+003

--- SQL operation complete.
>>explain options 'f' D8000;

LC   RC   OP   OPERATOR              OPT       DESCRIPTION           CARD
---- ---- ---- --------------------  --------  --------------------  ---------

7    .    8    root                            x                     1.40E+004
3    6    7    nested_join                                           1.40E+004
4    5    6    merge_union                                           2.00E+000
.    .    5    trafodion_vsbb_delet            INP002                1.00E+000
.    .    4    trafodion_vsbb_delet            INP001                1.00E+000
1    2    3    nested_join                                           7.00E+003
.    .    2    trafodion_delete                NP001                 1.00E+000
.    .    1    trafodion_scan                  NP001                 7.00E+003

--- SQL operation complete.
>>
>>-- initial condition
>>select count(*) from np001;

(EXPR)              
--------------------

               16384

--- 1 row(s) selected.
>>
>>execute D6000;

--- 5000 row(s) deleted.
>>display statistics;

Start Time             2015/10/07 20:26:18.466980
End Time               2015/10/07 20:26:49.052151
Elapsed Time                      00:00:30.585171
Compile Time                      00:00:00.120950
Execution Time                    00:00:30.585171


Table Name             Records        Records      Hdfs       Hdfs I/O    Hdfs Access
                      Accessed           Used      I/Os          Bytes     Time(usec)
TRAFODION.SEABASE.INP001
                          5000           5000         2         170000         482866
TRAFODION.SEABASE.NP001
                             0           5000        14         160000        5839776
TRAFODION.SEABASE.NP001
                             0           5000     10000         447428       21171910
TRAFODION.SEABASE.INP001
                             0           5000       820          46640        4949645
TRAFODION.SEABASE.INP002
                             0           5000       820          46640        5540131

--- SQL operation complete.
>>
>>-- restore the data to the table
>>insert into np001 select * from np001backup where b between 1000 and 6000;

--- 5000 row(s) inserted.
>>select count(*) from np001;

(EXPR)              
--------------------

               16384

--- 1 row(s) selected.
>>
>>execute D6500;

--- 5500 row(s) deleted.
>>display statistics;

Start Time             2015/10/07 20:27:07.270004
End Time               2015/10/07 20:27:29.752044
Elapsed Time                      00:00:22.482040
Compile Time                      00:00:00.109582
Execution Time                    00:00:22.482040


Table Name             Records        Records      Hdfs       Hdfs I/O    Hdfs Access
                      Accessed           Used      I/Os          Bytes     Time(usec)
TRAFODION.SEABASE.INP001
                          5500           5500         2         187000         196456
TRAFODION.SEABASE.NP001
                             0           5500        14         176000        2798004
TRAFODION.SEABASE.NP001
                             0           5500     11000         477250       16779875
TRAFODION.SEABASE.INP001
                             0           5500       917          51334        3559430
TRAFODION.SEABASE.INP002
                             0           5500       917          51334        3712274

--- SQL operation complete.
>>
>>-- restore the data to the table
>>insert into np001 select * from np001backup where b between 1000 and 6500;

--- 5500 row(s) inserted.
>>select count(*) from np001;

(EXPR)              
--------------------

               16384

--- 1 row(s) selected.
>>
>>execute D7000;

--- 6000 row(s) deleted.
>>display statistics;

Start Time             2015/10/07 20:27:45.187988
End Time               2015/10/07 20:28:12.012484
Elapsed Time                      00:00:26.824496
Compile Time                      00:00:00.150172
Execution Time                    00:00:26.824496


Table Name             Records        Records      Hdfs       Hdfs I/O    Hdfs Access
                      Accessed           Used      I/Os          Bytes     Time(usec)
TRAFODION.SEABASE.INP001
                          6000           6000         2         204000         201448
TRAFODION.SEABASE.NP001
                             0           6000        15         192000        2596367
TRAFODION.SEABASE.NP001
                             0           6000     12000         520500       20666863
TRAFODION.SEABASE.INP001
                             0           6000      1000          56000        4518636
TRAFODION.SEABASE.INP002
                             0           6000      1000          56000        4537803

--- SQL operation complete.
>>
>>-- restore the data to the table
>>insert into np001 select * from np001backup where b between 1000 and 7000;

--- 6000 row(s) inserted.
>>select count(*) from np001;

(EXPR)              
--------------------

               16384

--- 1 row(s) selected.
>>
>>execute D7500;

--- 6500 row(s) deleted.
>>display statistics;

Start Time             2015/10/07 20:28:30.778365
End Time               2015/10/07 20:28:59.217452
Elapsed Time                      00:00:28.439087
Compile Time                      00:00:00.113676
Execution Time                    00:00:28.439087


Table Name             Records        Records      Hdfs       Hdfs I/O    Hdfs Access
                      Accessed           Used      I/Os          Bytes     Time(usec)
TRAFODION.SEABASE.NP001
                         16384           6500         3         786432         378354
TRAFODION.SEABASE.NP001
                             0           6500     13000         563750       24031660
TRAFODION.SEABASE.INP001
                             0           6500      1083          60666        5218143
TRAFODION.SEABASE.INP002
                             0           6500      1083          60666        5363805

--- SQL operation complete.
>>
>>-- restore the data to the table
>>insert into np001 select * from np001backup where b between 1000 and 7500;

--- 6500 row(s) inserted.
>>select count(*) from np001;

(EXPR)              
--------------------

               16384

--- 1 row(s) selected.
>>
>>execute D8000;

--- 7000 row(s) deleted.
>>display statistics;

Start Time             2015/10/07 20:29:22.091740
End Time               2015/10/07 20:29:55.793977
Elapsed Time                      00:00:33.702237
Compile Time                      00:00:00.110113
Execution Time                    00:00:33.702237


Table Name             Records        Records      Hdfs       Hdfs I/O    Hdfs Access
                      Accessed           Used      I/Os          Bytes     Time(usec)
TRAFODION.SEABASE.NP001
                         16384           7000         3         786432         329139
TRAFODION.SEABASE.NP001
                             0           7000     14000         607000       28474991
TRAFODION.SEABASE.INP001
                             0           7000      1167          65334        5540065
TRAFODION.SEABASE.INP002
                             0           7000      1167          65334        5584084

--- SQL operation complete.
>>
>>-- restore the data to the table
>>insert into np001 select * from np001backup where b between 1000 and 8000;

--- 7000 row(s) inserted.
>>select count(*) from np001;

(EXPR)              
--------------------

               16384

--- 1 row(s) selected.
>>
>>
>>
>>
>>--obey TEST001(cleanupx);
>>exit;

End of MXCI Session

sureshsubbiah commented:
I am assuming this delivery is ready to be merged now. Will do so in a couple of hours unless there is feedback.

DaveBirdsall commented:
Yes, thank you, Suresh.
